dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,837 @@
+            return S_OK;
+        case SFVM_WINDOWCLOSING:
+            m_ShellViewWindow = NULL;
+            return S_OK;
+    }
+    return E_NOTIMPL;
+}
+
+IFACEMETHODIMP CCabFolder::GetIconOf(PCUITEMID_CHILD pidl, UINT flags, int *pIconIndex)
+{
+    if (CABITEM *item = CABITEM::Validate(pidl))
+    {
+        int index = MapPIDLToSystemImageListIndex(this, pidl, flags);
+        if (index == -1 && item->IsFolder())
+            index = (flags & GIL_OPENICON) ? SIID_FOLDEROPEN : SIID_FOLDER;
+        if (index != -1)
+        {
+            *pIconIndex = index;
+            return S_OK;
+        }
+    }
+    return S_FALSE;
+}
+
+static HRESULT GetFsPathFromIDList(PCIDLIST_ABSOLUTE pidl, PWSTR pszPath)
+{
+    BOOL ret = SHGetPathFromIDListW(pidl, pszPath);
+    if (!ret && ILIsEmpty(pidl))
+        ret = SHGetSpecialFolderPathW(NULL, pszPath, CSIDL_DESKTOPDIRECTORY, TRUE);
+    return ret ? S_OK : HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
+}
+
+static int CALLBACK FolderBrowseCallback(HWND hwnd, UINT uMsg, LPARAM lParam, LPARAM lpData)
+{
+    WCHAR buf[MAX_PATH];
+    switch (uMsg)
+    {
+        case BFFM_INITIALIZED:
+        {
+            if (LoadStringW(g_hInst, IDS_EXTRACT, buf, _countof(buf)))
+            {
+                // Remove leading and trailing dots
+                WCHAR *s = buf, *e = s + lstrlenW(s);
+                while (*s == '.') ++s;
+                while (e > buf && e[-1] == '.') *--e = UNICODE_NULL;
+                SendMessageW(hwnd, WM_SETTEXT, 0, (LPARAM)s);
+                SendMessageW(GetDlgItem(hwnd, IDOK), WM_SETTEXT, 0, (LPARAM)s);
+            }
+            if (lpData)
+            {
+                SendMessageW(hwnd, BFFM_SETEXPANDED, FALSE, lpData);
+                SendMessageW(hwnd, BFFM_SETSELECTION, FALSE, lpData);
+            }
+            break;
+        }
+
+        case BFFM_SELCHANGED:
+        {
+            SFGAOF wanted = SFGAO_FILESYSTEM | SFGAO_FOLDER, query = wanted | SFGAO_STREAM;
+            PCIDLIST_ABSOLUTE pidl = (PCIDLIST_ABSOLUTE)lParam;
+            BOOL enable = ILIsEmpty(pidl); // Allow the desktop
+            PCUITEMID_CHILD child;
+            IShellFolder *pSF;
+            if (SUCCEEDED(SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &pSF), &child)))
+            {
+                SFGAOF attrib = query;
+                if (SUCCEEDED(pSF->GetAttributesOf(1, &child, &attrib)))
+                    enable = (attrib & query) == wanted;
+                pSF->Release();
+            }
+            if (enable)
+            {
+                // We don't trust .zip folders, check the file-system to make sure
+                UINT attrib = SUCCEEDED(GetFsPathFromIDList(pidl, buf)) ? GetFileAttributes(buf) : 0;
+                enable = (attrib & FILE_ATTRIBUTE_DIRECTORY) && attrib != INVALID_FILE_ATTRIBUTES;
+            }
+            PostMessageW(hwnd, BFFM_ENABLEOK, 0, enable);
+            break;
+        }
+    }
+    return 0;
+}
+
+struct EXTRACTFILESDATA
```

Comment:
Considering this is c++,
you can use `CComPtr<IStream>` and the likes in this object, and `new` / `delete` it to have all those destructors do their cleanup